x/shield/keeper/purchase.go

```diff
@@ -137,94 +148,133 @@ func (k Keeper) IterateAllPurchases(ctx sdk.Context, callback func(purchase type
 func (k Keeper) RemoveExpiredPurchases(ctx sdk.Context) {
 	store := ctx.KVStore(k.storeKey)
 	iterator := k.PurchaseQueueIterator(ctx, ctx.BlockTime())
-
+	bondDenom := k.sk.BondDenom(ctx)
 	defer iterator.Close()
 	for ; iterator.Valid(); iterator.Next() {
-		var timeslice []types.PurchaseTxHash
+		var timeslice []types.PPPair
 		k.cdc.MustUnmarshalBinaryLengthPrefixed(iterator.Value(), &timeslice)
-		for _, entry := range timeslice {
-			purchase, _ := k.GetPurchase(ctx, entry)
-			pool, err := k.GetPool(ctx, purchase.PoolID)
-			if err == nil {
-				pool.Available = pool.Available.Add(purchase.Shield.AmountOf(k.sk.BondDenom(ctx)))
-				pool.Shield = pool.Shield.Sub(purchase.Shield)
-				k.SetPool(ctx, pool)
+		for _, ppPair := range timeslice {
+			purchaseList, _ := k.GetPurchaseList(ctx, ppPair.PoolID, ppPair.Purchaser)
+
+			for i := 0; i < len(purchaseList.Entries); i++ {
+				entry := purchaseList.Entries[i]
+				if entry.ExpirationTime.Before(ctx.BlockTime()) {
+					purchaseList.Entries = append(purchaseList.Entries[:i], purchaseList.Entries[i+1:]...)
+					i--
+					pool, err := k.GetPool(ctx, purchaseList.PoolID)
+					if err != nil {
+						panic(err)
+					}
+					pool.Available = pool.Available.Add(entry.Shield.AmountOf(bondDenom))
+					pool.Shield = pool.Shield.Sub(entry.Shield)
+					k.SetPool(ctx, pool)
+				}
 			}
-			k.DeletePurchase(ctx, purchase.TxHash)
+			if len(purchaseList.Entries) == 0 {
+				k.DeletePurchaseList(ctx, purchaseList.PoolID, purchaseList.Purchaser)
+				continue
```

Comment:
Here I actually prefer to eliminate the continue and just make a then/else pair, as the two branches are very simple and symmetric, also as the continue behavior is a natural implicit control flow here anyway.

Reply:
OK